plugins/blueprint.cpp

```diff
@@ -27,22 +29,63 @@
 #include "df/building_trapst.h"
 #include "df/building_water_wheelst.h"
 #include "df/building_workshopst.h"
+#include "df/world.h"
 
 using std::string;
 using std::endl;
 using std::vector;
 using std::ofstream;
-using std::swap;
-using std::find;
 using std::pair;
 using namespace DFHack;
-using namespace df::enums;
 
 DFHACK_PLUGIN("blueprint");
+REQUIRE_GLOBAL(world);
 
-enum phase {DIG=1, BUILD=2, PLACE=4, QUERY=8};
+struct blueprint_options {
+    // whether to display help
+    bool help = false;
 
-command_result blueprint(color_ostream &out, vector <string> &parameters);
+    // starting tile coordinate of the translation area (if not set then all
+    // coordinates are set to -30000)
+    df::coord start;
+
+    // dimensions of translation area. width and height are guaranteed to be
+    // greater than 0. depth can be positive or negative, but not zero.
+    int32_t width  = 0;
+    int32_t height = 0;
+    int32_t depth  = 0;
+
+    // base name to use for generated files
+    string name;
+
+    // whether to autodetect which phases to output
+    bool auto_phase = false;
+
+    // if not autodetecting, which phases to output
+    bool dig   = false;
+    bool build = false;
+    bool place = false;
+    bool query = false;
+
+    static struct_identity _identity;
+};
+static const struct_field_info blueprint_options_fields[] = {
```

Comment:
This is an interesting approach that I felt like I should comment on. This looks good for type-safety! Long-term, I think we should generate type definitions like these with codegen somehow instead of manually, since there could be failures down the road if these field definitions aren't kept in sync with the structure definition. (using `offsetof()` to calculate offsets should help guard against most issues... but as a counterpoint, I'm not sure if the field definitions need to be ordered by offset, so rearranging fields could potentially have unintended consequences, for example).

Am I correct in assuming that you didn't define this in [PluginStatics.cpp](https://github.com/DFHack/dfhack/blob/develop/library/PluginStatics.cpp) because the objects in question are short-lived and can't exist if the plugin is unloaded? (I was playing around with this and `opts` tended to end up at the same stack address every time!) I think this is safe, as long as you're sure that no Lua code retains references to `opts` (hopefully the garbage collector doesn't...). 

In the interests of avoiding duplication, you could also consider a macro like [`FLD()`](https://github.com/DFHack/dfhack/blob/41ff2c2e6c612814e09d03704f00bcdfb613db52/library/DataStaticsFields.cpp#L62), but the existing definition would admittedly be difficult to include.

Reply:
+1 for generating the `struct_field_info`. As far as I could tell, this is the first time a marshal-able type has been manually defined in DFHack, so I guess the need for a generator just never came up before.

You are correct as to why I didn't put this in `PluginStatics`. The stack-allocated options struct doesn't live beyond the `blueprint` command execution so I figured I'd keep all the definitions together in the `blueprint.cpp` file instead of gaining a consistent identity object, but splitting the object definitions into `PluginStatics`.

Yeah, I manually expanded the #defines that are used in the generated `struct_field_info`s when I was figuring out how to do this. Not using the existing (but inaccessible) macros felt dirty, but copying the macros into `blueprint.cpp` felt somehow dirtier : p  I didn't really want to get into refactoring `DataStatics*` in this PR, but that might be the way to go longer term.